From f2036d8bb7a23c13ec67e6e1dab141c4fdc30696 Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Wed, 23 Jul 2014 10:50:07 -0400
Subject: [virt] s390/sclp_vt220: Fix kernel panic due to early terminal input

Message-id: <1406112611-9862-20-git-send-email-brueckner@redhat.com>
Patchwork-id: 86334
O-Subject: [RHEL7.1 PATCH 19/23] [s390] s390/sclp_vt220: Fix kernel panic due to early terminal input
Bugzilla: 1088554
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Description
-----------
commit e106e4ea44699cc940a919924f2834b91f426d9f
Author: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Commit: Martin Schwidefsky <schwidefsky@de.ibm.com>

    s390/sclp_vt220: Fix kernel panic due to early terminal input

    A kernel panic might occur when there is terminal input available via
    the SCLP VT220 interface at an early time during the boot process.

    The processing of terminal input requires prior initialization which is
    done via an early_initcall function (init_workqueues) while the SCLP
    VT220 driver registers for terminal input during a console_initcall
    function (sclp_vt220_con_init). When there is terminal input available
    via the SCLP interface between console_initcall and early_initcall, a
    null pointer dereference occurs (system_wq is null).

    Fix this problem by moving the registration for terminal input to a
    device_initcall function (sclp_vt220_tty_init).

    Reported-by: Christian Borntraeger <borntraeger@de.ibm.com>
    Tested-by: Christian Borntraeger <borntraeger@de.ibm.com>
    Signed-off-by: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Bugzilla
--------
BZ 1088554
https://bugzilla.redhat.com/show_bug.cgi?id=1088554

Upstream status of the patch
----------------------------
The patch is upstream as of kernel version v3.15-rc2
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=e106e4ea44699cc940a919924f2834b91f426d9f

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7725759

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/s390/char/sclp_vt220.c b/drivers/s390/char/sclp_vt220.c
index 9a12db8..b9a9f72 100644
--- a/drivers/s390/char/sclp_vt220.c
+++ b/drivers/s390/char/sclp_vt220.c
@@ -97,13 +97,16 @@ static void sclp_vt220_pm_event_fn(struct sclp_register *reg,
 static int __sclp_vt220_emit(struct sclp_vt220_request *request);
 static void sclp_vt220_emit_current(void);
 
-/* Registration structure for our interest in SCLP event buffers */
+/* Registration structure for SCLP output event buffers */
 static struct sclp_register sclp_vt220_register = {
  .send_mask  = EVTYP_VT220MSG_MASK,
+ .pm_event_fn  = sclp_vt220_pm_event_fn,
+};
+
+/* Registration structure for SCLP input event buffers */
+static struct sclp_register sclp_vt220_register_input = {
  .receive_mask  = EVTYP_VT220MSG_MASK,
- .state_change_fn = NULL,
  .receiver_fn  = sclp_vt220_receiver_fn,
- .pm_event_fn  = sclp_vt220_pm_event_fn,
 };
 
 
@@ -715,9 +718,14 @@ static int __init sclp_vt220_tty_init(void)
  rc = tty_register_driver(driver);
  if (rc)
   goto out_init;
+ rc = sclp_register(&sclp_vt220_register_input);
+ if (rc)
+  goto out_reg;
  sclp_vt220_driver = driver;
  return 0;
 
+out_reg:
+ tty_unregister_driver(driver);
 out_init:
  __sclp_vt220_cleanup();
 out_driver:
-- 
1.7.1